### ISSUE TITLE
compiler crashes when passing subclass to superclass ref formal

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#5353]
### Summary of Problem

Compiler fails verification (with --verify) or generates illegal C code (with --no-verify).

### Steps to Reproduce

**Source Code:**

```chapel
class C    { }
class D: C { }

proc test(ref lhs: C) {}

var c: C;
var d: D;

test(c);
test(d); // must be an error
```

**Compile command:**
`chpl foo.chpl -o foo`

**Execution command:**
`./foo`

**Associated Future Test(s):**

test/functions/vass/subclass-to-ref-formal.future

#5352

### Configuration Information

- Output of `chpl --version`:
chpl Version 1.14.0.080fe1b

- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: none *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none *
CHPL_TASKS: fifo *
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: cstdlib *
CHPL_MAKE: gmake
CHPL_ATOMICS: intrinsics
CHPL_GMP: gmp
CHPL_HWLOC: none
CHPL_REGEXP: re2
CHPL_WIDE_POINTERS: struct
CHPL_AUX_FILESYS: none

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:

gcc 4.7.2

- (For Cray systems only) Output of `module list`:

n/a

**Comment by bradcray** [comment]
See also PR #5352

**Comment by vasslitvinov** [comment]
Correction: in the above code, `d.print();` is correct and actually works fine.

It is `test(d);` that's problematic due to the bug in question.

`d = 1.0;` exercises the same pattern as `test(d);`  (It runs successfully in my experiment.)

**Comment by noakesmichael** [comment]
The commentary in #5352 does a better job of describing the issue that is
being discussed here.  That commentary suggests:

class C    { }
class D: C { proc print { writeln(); } }

proc test(ref lhs: C) { lhs = new C(); }

var d = new D();   // Recognize that the static type of 'd' is D.
test(d);  // When this call is allowed, it is possible to replace 'd' with an instance of C.
d.print;  // passes static checking, invokes non-existing C.print at runtime

**Comment by benharsh** [comment]
@vasslitvinov : can this be closed now? Looks like the bug was resolved in #8812

**Comment by mppf** [comment]
I'm closing this issue because the code in the issue is giving an error as expected and the future was retired in PR #8812.
